### DISCUSSION TITLE
How to intercept console.logs from jsRuntime instance?

### DISCUSSION THREAD

**Discussion by LeoCaprile** [denoland/deno#18393] [Q&A] [answered]
So I follow some examples but when I print the returned value of the Global<Value>, I only get undefined.
There's a way to intercept the stdout of the execution of the script?
Thanks before hand.
```rust
// Initialize a runtime instance
  let mut runtime = JsRuntime::new(RuntimeOptions {
    ..Default::default()
  });

  runtime.execute_script("runtime", include_str!("./runtime/index.js")).unwrap();

  let res = runtime.execute_script("<anon>", "console.log('Hello World')").unwrap();
  let mut scope = &mut runtime.handle_scope();
  let value = res.open(&mut scope);
  let result = value.to_rust_string_lossy(&mut scope);

  println!("{result:?}");

```
terminal output:
![imagen](https://user-images.githubusercontent.com/71847541/227302164-37409109-f860-4387-bfd7-9e26b3485f3c.png)

**Comment by bartlomieju** [comment, marked as answer]
I'm suprised you get "Hello World" printed at all. `deno_core` has only access to the `console` object provided by V8, which by default doesn't print at all (though it redirects the logs to the inspector if it is connected).

The `undefined` is printed because the value returned from evaluation of `console.log()` is `undefined`.

If you want to have more control over console, you should use `deno_console` crate, you can then create a new `console` instance like this:
```
globalThis.console = new console.Console((msg, level) => Deno.core.print(msg, level > 1));
```

The argument that the constructor accepts is a "print" function that in the case above uses built-in `Deno.core.print` function to output to stdout/stderr.

**Reply by LeoCaprile** [reply]
Thanks for your answer!

> I'm suprised you get "Hello World" printed at all. deno_core has only access to the console object provided by V8, which by default doesn't print at all (though it redirects the logs to the inspector if it is connected).

   I actually made my own implementation of console using Deno.core.print in line 6 I import the script to do it

> If you want to have more control over console, you should use deno_console crate, you can then create a new console instance like this:

   Great! didn't know about this, so to get the console.logs results of the executed jsRuntime and be able to return it from a function as a String for example.

```rust
fn returned_logs -> String {
  //...code
return logs
}
```
 I had to connect to the inspector or can I use the crate deno_console to be able to return the console.logs as a String?

**Reply by bartlomieju** [reply]
@LeoCaprile if you want your `console.log` to return string as a return value you'd do something like this:
```js
globalThis.console = {
  log: (msg) => {
    Deno.core.print(msg);
    return msg;
  }
}
```
Then in Rust:
```rust
let result js_runtime.execute_script("my_script", "console.log('hello')"); // <---- notice there's no trailing `;` after `console.log`
let val = result.unwrap();
let val_str = val.to_rust_string_lossy(scope);
```
